### PR TITLE
Added basic functionality for url encoding

### PR DESCRIPTION
**Pull request by ben-jones**
Created a urlEncode module with a variety of methods for url encoding and a series of unit tests to validate operation. Currently, cookie encoding does not work, but url encoding does.

### CODE REVIEW

File: code/urlEncode.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  data = matches.group('hash')
+  # strip any padding
+  data = data.rstrip('ABCDEF')
+  data = binascii.unhexlify(data)
+  return data
+
+def decode(data):
+  """Decode the given data after matching the url hiding format
+
+  Parameters: data- the url and cookies to be decoded in the form of a
```
Comment (gsathya) [inline code/urlEncode.py:193]: This is mildly confusing. In the previous functions, data is just a string but now data is a dict in this function. Maybe rename data?

File: code/urlEncode.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+def encodeAsCookie(data):
+  """Hide data inside a cookie"""
+  #todo fill out this functino
+  return
+
+def pickDomain():
+  """Pick a random domain from the list"""
+  domains = ['live.com', 'microsoft.com', 'baidu.com', 'hao123.com']
+  return domains[randint(0, len(domains)-1)]
```
Comment (gsathya) [inline code/urlEncode.py]:
``` python
from random import choice
return choice(domains)
```

File: code/urlEncode.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    return encodeAsBaidu(data)
+  elif encodingType == 'google':
+    return encodeAsGoogle(data)
+
+def encodeAsCookies(data):
+  """Hide data inside a series of cookies"""
+
+  cookies = []
+  while data != '':
+      encodeAsCookie(data[:BYTES_PER_COOKIE])
```
Comment (gsathya) [inline code/urlEncode.py]: check if data has enough bytes first, otherwise it might lead to a `IndexError: string index out of range`

File: code/urlEncode.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  Returns: a string with the decoded data
+
+  """
+  
+  pattern = 'click.*\?qs=(?P<hash>[0-9a-fA-F]*)'
+  matches = re.search(pattern, url)
+  data = matches.group('hash')
+  # strip any padding
+  data = data.rstrip('ABCDEF')
```
Comment (gsathya) [inline code/urlEncode.py]: what about padding on the left side? `data.strip` -- I'm not if there will be padding on the left side, but just making sure.

File: code/urlEncode.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  Parameters: data- the url and cookies to be decoded in the form of a
+  dictionary with the url stored under the key 'url' and an array of 0
+  or more cookies stored under 'cookie'
+
+  Returns: the decoded data in the form of a string
+
+  """
+  url = data['url']
+  cookies = data['cookie']
+  retVal = ''
```
Comment (gsathya) [inline code/urlEncode.py]: It's more efficient to have retVal as a list than a string. Strings are immutable in python, so every time you append data to it, it gets created again. Just do `retVal.append(decodeAsMarket(url))` and in the end do `return ''.join(retVal)`. Also, maybe pick a more expressive variable name than retVal?

File: code/urlEncode.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  cookies if data was stored in cookies.
+
+  Note: Cookies will be used if more data is sent than the url
+  encoding method can support. When this occurs, as many cookies as
+  necessary will be used to encode the data. Please be mindful that
+  large numbers of cookies will look suspicious
+
+  """
+  if encodingType not in AVAILABLE_TYPES:
+      raise(UrlEncodeError("Bad encoding type. Please refer to url-encode.AVAILABLE_TYPES for available options"))
```
Comment (gsathya) [inline code/urlEncode.py]: Line wrap please

File: code/urlEncode.py (hunk 4 above)
Comment (ben-jones) [inline code/urlEncode.py]: There should only be padding on the right, but I changed it anyway.

File: code/verifyUrlEncode.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    """Create random 80 character strings of hex for testing"""
+
+    characters = ['0','1','2','3','4','5','6','7',
+                  '8','9','a','b','c','d','e','f']
+    index = 0
+    stringy = ''
+    while index < 80:
+      stringy += characters[randint(0,15)]
+      index += 1
+    return stringy
```
Comment (gsathya) [inline code/verifyUrlEncode.py]: lol stringy? :P

File: code/verifyUrlEncode.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  def test_decodeAsMarket(self):
+    """Verify that decodeAsMarket correctly decodes data"""
+
+    testData = [ self.gen_random_hex_chars() for index in range(5)]
+    testData[4] = testData[4][:76]
+    testStrings = [binascii.unhexlify(testData[index]) for index in range(5)]
+    #include a test to be sure that we can correctly decode padding
+    testData[4] += "ABCD"
+    for datum in testData:
+      url = 'click.live.com' + '?qs=' + datum
```
Comment (gsathya) [inline code/verifyUrlEncode.py]: yeah, don't do this. Check out http://docs.python.org/2/library/urllib.html#examples

File: code/verifyUrlEncode.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+      self.assertTrue(urlEncode.isMarket(datum))
+    for datum in falseData:
+      self.assertFalse(urlEncode.isMarket(datum))
+
+  def test_decodeAsMarket(self):
+    """Verify that decodeAsMarket correctly decodes data"""
+
+    testData = [ self.gen_random_hex_chars() for index in range(5)]
+    testData[4] = testData[4][:76]
+    testStrings = [binascii.unhexlify(testData[index]) for index in range(5)]
```
Comment (gsathya) [inline code/verifyUrlEncode.py]: why not do -

```
[binascii.unhexlify(datum) for datum in testData]
```

File: code/verifyUrlEncode.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    for datum in trueData:
+      self.assertTrue(urlEncode.isMarket(datum))
+    for datum in falseData:
+      self.assertFalse(urlEncode.isMarket(datum))
+
+  def test_decodeAsMarket(self):
+    """Verify that decodeAsMarket correctly decodes data"""
+
+    testData = [ self.gen_random_hex_chars() for index in range(5)]
+    testData[4] = testData[4][:76]
```
Comment (gsathya) [inline code/verifyUrlEncode.py:152]: Why 76? Sounds like a magic number ;)

File: code/verifyUrlEncode.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+      char = urlEncode.pickRandomHexChar()
+      self.assertIn(char, characters)
+
+  def test_encodeAsMarket(self):
+    """Verify that data are correctly stored in the url market form"""
+    
+    testData = ['some text', 'more text', 
+                'even more text that seems moderately lon']
+    for datum in testData:
+        output = urlEncode.encodeAsMarket(datum)
```
Comment (gsathya) [inline code/verifyUrlEncode.py:120]: output? what is output? Maybe use decodedUrl?

File: code/verifyUrlEncode.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+      self.assertEqual(testOutput, testString)
+
+  def gen_random_hex_chars(self):
+    """Create random 80 character strings of hex for testing"""
+
+    characters = ['0','1','2','3','4','5','6','7',
+                  '8','9','a','b','c','d','e','f']
+    index = 0
+    stringy = ''
+    while index < 80:
```
Comment (gsathya) [inline code/verifyUrlEncode.py]: Why use a while?

```
random_hex = []
for x in range(80):
  random_hex.append(choice(characters))
return ''.join(random_hex)
```

File: code/urlEncode.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+def encodeAsCookie(data):
+  """Hide data inside a cookie"""
+  #todo fill out this functino
+  return
+
+def pickDomain():
+  """Pick a random domain from the list"""
+  domains = ['live.com', 'microsoft.com', 'baidu.com', 'hao123.com']
+  return choice(domains)
```
Comment (gsathya) [inline code/urlEncode.py:127]: use `choice` instead of randint

File: code/verifyUrlEncode.py (hunk 10 above)
Comment (ben-jones) [inline code/verifyUrlEncode.py:152]: This was a just a random number that I picked to perform this test.

**Comment by ben-jones** [comment]
I chose not to inline the functionality of the test for 2 reasons. 1) I wanted to be sure that I was the test was not dependent on another function and 2) I wanted to add flexibility to the unittest. If we suddenly think of an edge case, it is trivial to add it now. Ultimately, this is just personal preference though.

**Comment by gsathya** [comment]
> I chose not to inline the functionality of the test for 2 reasons. 1) I wanted to be sure that I was the test was not dependent on another function and 2) I wanted to add flexibility to the unittest. If we suddenly think of an edge case, it is trivial to add it now. Ultimately, this is just personal preference though.

No no, in `self.assertEqual(urlEncode.decodeAsCookie(cookie), datum)`, you're deriving the value for `cookie` exactly as how you would in encodeAsCookie. Shouldn't you just hardcode the value of `cookie`? Like for ex  `self.assertEqual(urlEncode.decodeAsCookie("some_cookie_value_that_is_hardcoded", datum)`

**Comment by gsathya** [comment]
2dbf1bc990ecb8076ad1938c88885f43a1002875 needs the second line in the commit message to be line wrapped. I suggest using M-x fill-paragraph.

Amend the commit message and force push it. I'll merge it.

**Comment by gsathya** [comment]
> Amend the commit message and force push it. I'll merge it.

I've fixed it.

**Comment by gsathya** [comment]
Merged. Thanks!
